File: e-java/src/test/java/e/java/test/Assertions.java
package e.java.test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import e.java.E;
import e.java.EOr;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class Assertions {
    private Assertions() {}

    public static void assertCode(E e, Integer code) {
        assertEquals(code != null,              e.hasCode());
        assertEquals(Optional.ofNullable(code), e.code());
    }

    public static void assertName(E e, String name) {
        assertEquals(name != null,              e.hasName());
        assertEquals(Optional.ofNullable(name), e.name());
    }

    public static void assertMessage(E e, String message) {
        assertEquals(message != null,              e.hasMessage());
        assertEquals(Optional.ofNullable(message), e.message());
    }

    public static void assertCauses(E e, List<E> causes) {
        assertEquals(!causes.isEmpty(), e.hasCause());
        assertEquals(causes,            e.causes());
    }

    public static void assertData(E e, Map<String, String> data) {
        assertEquals(!data.isEmpty(), e.hasData());
        assertEquals(data,            e.data());
    }

    public static void assertTime(E e, Long time) {
        assertEquals(time != null,              e.hasTime());
        assertEquals(Optional.ofNullable(time), e.time());
    }

    public static void assertAlmostSame(Long expected, Long actual, Long threshold) {
        assertTrue(
            Math.abs(actual - expected) < threshold,
            String.format(
                "Times were apart mode than %d ms, expected: %d, actual: %d",
                threshold,
                expected,
                actual
            )
        );
    }

    public static void assertAlmostSame(Long expected, Long actual) {
        assertAlmostSame(expected, actual, 1000L);
    }

    public static <A> void assertError(EOr<A> eor, E thatE) {
        eor.fold(
            thisE -> { assertEquals(thatE, thisE, "Expected: " + thisE.trace() + ", Actual: "+ thatE.trace()); return null; },
            a     -> fail(String.format("EOr did not contain error, it contained: %s", a))
        );
    }

    public static <A> void assertValue(EOr<A> eor, A thatA) {
        eor.fold(
            e     -> fail(String.format("EOr did not contain value, it contained: %s", e.trace())),
            thisA -> { assertEquals(thatA, thisA); return null; }
        );
    }
}

File: e-kotlin/src/test/kotlin/e/kotlin/test/Assertions.kt
package e.kotlin.test

import e.kotlin.E
import e.kotlin.EOr
import org.junit.jupiter.api.Assertions.*
import kotlin.math.absoluteValue

interface Assertions {
    fun E.assertCode(code: Int?) {
        assertEquals(code != null, this.hasCode)
        assertEquals(code,         this.code)
    }

    fun E.assertName(name: String?) {
        assertEquals(name != null, this.hasName)
        assertEquals(name,         this.name)
    }

    fun E.assertMessage(message: String?) {
        assertEquals(message != null, this.hasMessage)
        assertEquals(message,         this.message)
    }

    fun E.assertCauses(causes: List<E>) {
        assertEquals(causes.isNotEmpty(), this.hasCause)
        assertEquals(causes,              this.causes)
    }

    fun E.assertData(data: Map<String, String>) {
        assertEquals(data.isNotEmpty(), this.hasData)
        assertEquals(data,              this.data)
    }

    fun E.assertTime(time: Long?) {
        assertEquals(time != null, this.hasTime)
        assertEquals(time,         this.time)
    }

    fun assertAlmostSame(expected: Long, actual: Long?, threshold: Long = 1000L) {
        assertTrue(
            ((actual ?: 0L) - expected).absoluteValue < threshold,
            "Times were apart mode than $threshold ms, expected: $expected, actual: $actual"
        )
    }

    fun <A> EOr<A>.assertError(thatE: E) {
        fold(
            { thisE -> assertEquals(thatE, thisE, "Expected: ${thatE.trace()}, Actual: ${thisE.trace()}") },
            { a     -> fail<A>("EOr did not contain error, it contained: $a") }
        )
    }

    fun <A> EOr<A>.assertValue(thatA: A) {
        fold(
            { e     -> fail<A>("EOr did not contain value, it contained: ${e.trace()}") },
            { thisA -> assertEquals(thatA, thisA) }
        )
    }
}
